### PR TITLE
add deadline timer to outgoing http sessions

### PR DESCRIPTION
**Pull request by msgmaxim**
Also improved encapsulation in HttpClientSession

### CODE REVIEW

File: httpserver/http_connection.h [hunk 1]
```diff
@@ -50,27 +50,32 @@ class HttpClientSession
...
     void on_read(boost::system::error_code ec, std::size_t bytes_transferred);
 
     void init_callback(std::shared_ptr<std::string>&& body);
 
   public:
-    tcp::socket socket_;
     // Resolver and socket require an io_context
-    explicit HttpClientSession(boost::asio::io_context& ioc,
-                               const request_t& req, http_callback_t cb);
+    HttpClientSession(boost::asio::io_context& ioc, const tcp::endpoint& ep,
```
Comment (sachaaaaa) [inline httpserver/http_connection.h]: Just curious why `explicit` was removed?

Comment (msgmaxim) [reply httpserver/http_connection.h]: It is only necessary for constructors that take 1 argument, so you couldn't do `HttpClientSession = ioc;`. Now that we have many arguments it is not needed.

Comment (sachaaaaa) [reply httpserver/http_connection.h]: It still prevents doing `HttpClientSession session = {ioc, endpoint, ...}` tho right?

Comment (msgmaxim) [reply httpserver/http_connection.h]: I don't think you can do that with or without the `explicit` keyword.

Comment (sachaaaaa) [reply httpserver/http_connection.h]: Ok, let's keep the removal of explicit :)

Comment (msgmaxim) [reply httpserver/http_connection.h]: Just to summarise, it turns out it did prevent `HttpClientSession session = {ioc, endpoint, ...}`, but that shouldn't be a problem anyway.

**Pull request merged**